### DISCUSSION TITLE
Lines of code: Exempt first commit?

### DISCUSSION THREAD

**Discussion by ctjlewis** [lowlighter/metrics#1109] [Ideas]
Hey Simon, hope work on this tool has been going well.

I’ve been looking at my SLOC count and was wondering what it would be if initial commits were exempted, since that’s a more accurate representation of actual lines written.

Can you point me to the relevant part of the codebase where I might maybe add this flag, if I wanted to PR (and would you entertain a merge)?

**Comment by lowlighter** [comment]
Hi 👋 !

If you're talking about the `line` plugin, I don't think it's directly possible as we use [`octokit.rest.repos.getContributorsStats`](https://octokit.github.io/rest.js/v18#repos-get-contributors-stats)

However it should be possible to fetch the first commit of each repositories through the GitHub API and get the diff stats.
From there just substracting the result from the computed total should give you numbers similar to what you want to achieve

The revelant code is here:
https://github.com/lowlighter/metrics/blob/79b80b1900ef8f2b3abbaad91151276327d9278a/source/plugins/lines/index.mjs#L23-L38

To add a new option, you just need to patch the revelant [`metadata.yml`](https://github.com/lowlighter/metrics/blob/master/source/plugins/lines/metadata.yml)

And then you can extract it through `imports.metadata.plugins.lines.inputs`, which will automatically perform input sanitization and casting:
https://github.com/lowlighter/metrics/blob/79b80b1900ef8f2b3abbaad91151276327d9278a/source/plugins/lines/index.mjs#L10

If you're satisfied with the result and feel like it'd be a great addition, feel free to open a PR !

**Comment by jayvdb** [comment]
@ctjlewis, out of curiousity, could you clarify why excluding initial commits would be a more accurate representation of actual lines written?

I see you have "12.5m added, 4.34m removed" which is a mean feat, and needs some revising downwards to be believable unless you've been hammering out code constantly for 50 years, and somehow miraculously GitHub has it all with your authorship on it.

I am also skeptical of my "860k added", but my "17.3k removed" is a bit more believable.

It seems the plugin is iterating over the users repositories - so I wonder if that is including forks as well - i.e. is the same/similar commit in a fork and the upstream being counted, which would be a quick way of inflating the numbers, but only by double, which probably isnt enough to explain your "12.5m added".

It would be nice to see the lines plugin have a table/chart mode, so we can see which repos are at the top - they are likely where we will find the inflation occurring, and maybe this could then be remedied by using adding skip parameters.

Could it be that your numbers include a lot of generated JS derived (multiple times for different JS flavours) from the TS which you authored?

And/Or is this forks of another project which you've radically revised as your first commit in a repo which stretches back before you?

It looks like [`octokit.rest.repos.getContributorsStats`](https://octokit.github.io/rest.js/v18#repos-get-contributors-stats) works on weeks, so you might be able to do a quick hack to remove your first "week" of contributions to each repository.

**Reply by ctjlewis** [reply]
My impression was many lines are initial commits from templates, and yarn.lock files as well. There's just no real sane way to exclude them.

No generated JS though: TS build output isn't checked into git typically.

**Reply by ctjlewis** [reply]
What's also weird is it doesn't seem to be driven by forks. I've seen it go up as I merge more and more PRs but I don't have very many forked repos. 

It's really just asking Github what the additions and deletions per week are, over the lifetime of your account, and summing them. I will probably run the count myself and see if anything stands out.

I'm curious about the JS output question now though, because my languages stats still indicate like 5:1 JS:TS when I've mostly written TS.

**Reply by jayvdb** [reply]
Important to note that the API being used is per repo - https://docs.github.com/en/rest/metrics/statistics#get-all-contributor-commit-activity

So you could manually query via the API a few likely repos, and should be able to see unexpected inflation happening.

**Reply by lowlighter** [reply]
Maybe we could also add a setting to display the "top repo" with their diff, I could see it interesting, and it could also helps understand why these stats may be that high ?

**Reply by ctjlewis** [reply]
@lowlighter Anything we could do to get the process to print out each repo’s stats as it loops over them would be fine to start, cause then we could just pull the log and scan real quick.

**Reply by lowlighter** [reply]
@ctjlewis  

Sorry for the format:
```
metrics/compute/ctjlewis/plugins > lines > querying api
metrics/compute/ctjlewis/plugins > lines > computing total diff
metrics/compute/ctjlewis/plugins > lines > ctjlewis/web-widgets-primer: @ctjlewis +614 -21 ~7
metrics/compute/ctjlewis/plugins > lines > ctjlewis/web-widgets-poc: @ctjlewis +142 -1 ~2
metrics/compute/ctjlewis/plugins > lines > ctjlewis/typewriter: @ctjlewis +11707 -3704 ~73
metrics/compute/ctjlewis/plugins > lines > ctjlewis/enable-window-document: @ctjlewis +848 -88 ~34
metrics/compute/ctjlewis/plugins > lines > ctjlewis/enable-browser-mode: @ctjlewis +25266 -24374 ~86
metrics/compute/ctjlewis/plugins > lines > ctjlewis/state-machine: @ctjlewis +3884 -974 ~12
metrics/compute/ctjlewis/plugins > lines > ctjlewis/web-widgets: @ctjlewis +26587 -22876 ~155
metrics/compute/ctjlewis/plugins > lines > ctjlewis/react-spotify-embed: @ctjlewis +14048 -2058 ~19
metrics/compute/ctjlewis/plugins > lines > ctjlewis/wiktionary-viewer: @ctjlewis +4924 -270 ~13
metrics/compute/ctjlewis/plugins > lines > ctjlewis/next-endpoint: @ctjlewis +12479 -8042 ~33
metrics/compute/ctjlewis/plugins > lines > ctjlewis/iframe-pointer-events: @ctjlewis +3826 -0 ~1
metrics/compute/ctjlewis/plugins > lines > ctjlewis/bionic-reading-uwu: @ctjlewis +2018 -11 ~2
metrics/compute/ctjlewis/plugins > lines > ctjlewis/create-esnext: @ctjlewis +2218 -51 ~7
metrics/compute/ctjlewis/plugins > lines > ctjlewis/poker-combinatorics: @ctjlewis +319 -4 ~4
metrics/compute/ctjlewis/plugins > lines > ctjlewis/await-shell: @ctjlewis +6222 -4246 ~131
metrics/compute/ctjlewis/plugins > lines > ctjlewis/css-truncate-text: @ctjlewis +4048 -4 ~4
metrics/compute/ctjlewis/plugins > lines > ctjlewis/create-scoped-state: @ctjlewis +15788 -3807 ~19
metrics/compute/ctjlewis/plugins > lines > ctjlewis/create-debug-logger: @ctjlewis +2811 -381 ~29
metrics/compute/ctjlewis/plugins > lines > ctjlewis/react-library-starter: @ctjlewis +12416 -2829 ~18
metrics/compute/ctjlewis/plugins > lines > ctjlewis/react-user-flow: @ctjlewis +9619 -38 ~2
metrics/compute/ctjlewis/plugins > lines > ctjlewis/ctjlewis: @ctjlewis +244 -85 ~55
metrics/compute/ctjlewis/plugins > lines > ctjlewis/tailwind-recursion-bug-repro: @ctjlewis +4102 -550 ~2
metrics/compute/ctjlewis/plugins > lines > ctjlewis/next-tailwind-starter: @ctjlewis +5798 -3086 ~10
metrics/compute/ctjlewis/plugins > lines > ctjlewis/next-spa: @ctjlewis +8910 -12 ~4
metrics/compute/ctjlewis/plugins > lines > ctjlewis/toy-problems-are-bad: @ctjlewis +7593 -1 ~3
metrics/compute/ctjlewis/plugins > lines > ctjlewis/rollup-node-testing: @ctjlewis +11595409 -3448492 ~20
metrics/compute/ctjlewis/plugins > lines > ctjlewis/react-components-library-template: @ctjlewis +7520 -250 ~28
metrics/compute/ctjlewis/plugins > lines > ctjlewis/social-authentication-notes: @ctjlewis +174 -0 ~1
metrics/compute/ctjlewis/plugins > lines > ctjlewis/remote-design-tokens: @ctjlewis +7873 -84 ~9
metrics/compute/ctjlewis/plugins > lines > ctjlewis/remote-styles: @ctjlewis +7 -1 ~2
metrics/compute/ctjlewis/plugins > lines > ctjlewis/Levels: @ctjlewis +460 -0 ~1
metrics/compute/ctjlewis/plugins > lines > ctjlewis/FinGraph: @ctjlewis +4025 -29 ~6
metrics/compute/ctjlewis/plugins > lines > ctjlewis/next-crypto-icons: @ctjlewis +11436 -34 ~10
metrics/compute/ctjlewis/plugins > lines > ctjlewis/default-resolution-bug: @ctjlewis +3275 -0 ~2
metrics/compute/ctjlewis/plugins > lines > ctjlewis/next-ses: @ctjlewis +3771 -136 ~2
metrics/compute/ctjlewis/plugins > lines > ctjlewis/next-esm-re-exports: @ctjlewis +18525 -5615 ~6
metrics/compute/ctjlewis/plugins > lines > ctjlewis/stateful-component: @ctjlewis +33510 -21611 ~39
metrics/compute/ctjlewis/plugins > lines > ctjlewis/export-ts: @ctjlewis +14471 -13787 ~41
metrics/compute/ctjlewis/plugins > lines > ctjlewis/blog: @ctjlewis +82057 -59787 ~20
metrics/compute/ctjlewis/plugins > lines > ctjlewis/shebang-repro: @ctjlewis +3599 -53 ~3
metrics/compute/ctjlewis/plugins > lines > ctjlewis/rn-disk-cache: @ctjlewis +16618 -5889 ~28
metrics/compute/ctjlewis/plugins > lines > ctjlewis/rainbow-components: @ctjlewis +14638 -463 ~4
metrics/compute/ctjlewis/plugins > lines > ctjlewis/rainbow-overrides: @ctjlewis +23821 -14224 ~147
metrics/compute/ctjlewis/plugins > lines > ctjlewis/replace: @ctjlewis +1527 -0 ~1
metrics/compute/ctjlewis/plugins > lines > ctjlewis/BitcloutOSINT: @ctjlewis +3235 -0 ~1
metrics/compute/ctjlewis/plugins > lines > ctjlewis/rainbow-token-service: @ctjlewis +3985 -516 ~12
metrics/compute/ctjlewis/plugins > lines > ctjlewis/perf-track-analysis: @ctjlewis +410 -88 ~3
metrics/compute/ctjlewis/plugins > lines > ctjlewis/gql-ui-tsdx: @ctjlewis +15599 -44 ~5
metrics/compute/ctjlewis/plugins > lines > ctjlewis/gql-ui: @ctjlewis +7255 -125 ~2
metrics/compute/ctjlewis/plugins > lines > ctjlewis/amplify-trigger-repro: @ctjlewis +8505 -4443 ~13
metrics/compute/ctjlewis/plugins > lines > ctjlewis/amplify-dev: @ctjlewis +8442 -4002 ~10
metrics/compute/ctjlewis/plugins > lines > ctjlewis/webpack-add-styles-to-document: @ctjlewis +628 -0 ~1
metrics/compute/ctjlewis/plugins > lines > ctjlewis/next-typescript-tailwind: @ctjlewis +2833 -0 ~1
metrics/compute/ctjlewis/plugins > lines > ctjlewis/next-typescript: @ctjlewis +2227 -0 ~1
metrics/compute/ctjlewis/plugins > lines > ctjlewis/netlify-fauna-typescript-example: @ctjlewis +15838 -9909 ~2
metrics/compute/ctjlewis/plugins > lines > ctjlewis/object-switch: @ctjlewis +2043 -752 ~3
metrics/compute/ctjlewis/plugins > lines > ctjlewis/christiantjl: @ctjlewis +13 -1 ~2
metrics/compute/ctjlewis/plugins > lines > ctjlewis/unicode-category-ranges: @ctjlewis +393 -0 ~2
metrics/compute/ctjlewis/plugins > lines > ctjlewis/export_repro: @ctjlewis +401 -3 ~3
metrics/compute/ctjlewis/plugins > lines > ctjlewis/closure-define-repro: @ctjlewis +69 -21 ~4
metrics/compute/ctjlewis/plugins > lines > ctjlewis/JSDOM-debugging: @ctjlewis +965 -0 ~1
metrics/compute/ctjlewis/plugins > lines > ctjlewis/closure-compiler-modules: @ctjlewis +663 -13 ~2
metrics/compute/ctjlewis/plugins > lines > ctjlewis/login: @ctjlewis +2365 -232 ~9
metrics/compute/ctjlewis/plugins > lines > ctjlewis/find: @ctjlewis +5800 -239 ~14
metrics/compute/ctjlewis/plugins > lines > ctjlewis/app: @ctjlewis +323 -17 ~4
metrics/compute/ctjlewis/plugins > lines > ctjlewis/directional_layout: @ctjlewis +660 -47 ~5
metrics/compute/ctjlewis/plugins > lines > completed
```

Seems to be `ctjlewis/rollup-node-testing: @ctjlewis +11595409 -3448492 ~20` the culprit

___

Used this patch, will probably merge it to main branch when trying to integrate the lines detailed stats per repo
```diff
diff --git a/source/plugins/lines/index.mjs b/source/plugins/lines/index.mjs
index 1fd00b43..850d651f 100644
--- a/source/plugins/lines/index.mjs
+++ b/source/plugins/lines/index.mjs
@@ -23,18 +23,24 @@ export default async function({login, data, imports, rest, q, account}, {enabled
     //Get contributors stats from repositories
     console.debug(`metrics/compute/${login}/plugins > lines > querying api`)
     const lines = {added: 0, deleted: 0, changed: 0}
-    const response = [...await Promise.allSettled(repositories.map(({repo, owner}) => (skipped.includes(repo.toLocaleLowerCase())) || (skipped.includes(`${owner}/${repo}`.toLocaleLowerCase())) ? {} : rest.repos.getContributorsStats({owner, repo})))].filter(({status}) => status === "fulfilled").map(({value}) => value)
+    const response = [...await Promise.allSettled(repositories.map(async ({repo, owner}) => (skipped.includes(repo.toLocaleLowerCase())) || (skipped.includes(`${owner}/${repo}`.toLocaleLowerCase())) ? {} : {handle:`${owner}/${repo}`, stats:(await rest.repos.getContributorsStats({owner, repo})).data}))].filter(({status}) => status === "fulfilled").map(({value}) => value)

     //Compute changed lines
     console.debug(`metrics/compute/${login}/plugins > lines > computing total diff`)
-    response.map(({data: repository}) => {
+    response.map(({handle, stats}) => {
       //Check if data are available
-      if (!Array.isArray(repository))
+      if (!Array.isArray(stats))
         return
       //Compute editions
-      const contributors = repository.filter(({author}) => context.mode === "repository" ? true : author?.login?.toLocaleLowerCase() === login.toLocaleLowerCase())
-      for (const contributor of contributors)
-        contributor.weeks.forEach(({a = 0, d = 0, c = 0}) => (lines.added += a, lines.deleted += d, lines.changed += c))
+      const contributors = stats.filter(({author}) => context.mode === "repository" ? true : author?.login?.toLocaleLowerCase() === login.toLocaleLowerCase())
+      for (const contributor of contributors) {
+        let added = 0, deleted = 0, changed = 0
+        contributor.weeks.forEach(({a = 0, d = 0, c = 0}) => (added += a, deleted += d, changed += c))
+        console.debug(`metrics/compute/${login}/plugins > lines > ${handle}: @${contributor.author.login} +${added} -${deleted} ~${changed}`)
+        lines.added += added
+        lines.deleted += deleted
+        lines.changed += changed
+      }
     })

     //Results
```

**Reply by ctjlewis** [reply]
No worries. I can certainly see 11M of the lines in rollup-node-testing, presumably from a huge yarn.lock plus all the testing tools. 

A lot of it does seem to be `yarn.lock` files making the smallest # lines changed over a new repo and several package updates no smaller than 3,000 lines.

There is no way for us to know anything about which files the lines changed for a repo without adding really complicated, unrewarding logic. Ultimately I will just call this “code/lines moved/pushed” or something, rather than “code/lines written,” as it does depict the total number of lines of code I have modified in GitHub (even if not directly written).

Comparatively, between two users, this effect will hit everyone equally and so the metrics are not really distorted. Though it’s clear I did not actually *write* 12M lines, that is how much code I’ve moved.

**Reply by ctjlewis** [reply]
Lmao. Next time I’m asked for a resume I’m just gonna send the text dump of this and a link to this discussion.

**Reply by ctjlewis** [reply]
Also, for that log - it could be awesome to add a log in the weeks loop, and print out the progress for every week. Kind of insane to see your entire GitHub history print out week by week, but it’s one small tweak from that.

Also, thank you for helping on this too Simon. Both times I’ve asked about using this in a new way you’ve been very helpful and generous with your time.

**Reply by lowlighter** [reply]
> Also, for that log - it could be awesome to add a log in the weeks loop, and print out the progress for every week. Kind of insane to see your entire GitHub history print out week by week, but it’s one small tweak from that.

Will eventually be implemented by #1151
While it'll default to current year, it can be set to whole history. The scaling is really rough though (linear scale) so if there are hight spikes then the whole graph may look squeezed atm
Maybe will try to check if logscale gives better result in a future pr
